File: src/main/java/org/moinex/util/APIUtils.java
/*
 * Filename: APIUtils.java
 * Created on: January 17, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.util;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.PosixFilePermissions;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import org.json.JSONObject;
import org.moinex.error.MoinexException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Utility class for managing API-related operations, like fetching stock prices and
 * market indicators
 */
public class APIUtils {
    private static final ExecutorService executorService = Executors.newCachedThreadPool();

    private static final List<Process> runningProcesses = new ArrayList<>();
    private static final Logger logger = LoggerFactory.getLogger(APIUtils.class);
    private static boolean shuttingDown = false;

    // Prevent instantiation
    private APIUtils() {}

    /**
     * Shutdown the executor service
     */
    public static synchronized void shutdownExecutor() {
        if (shuttingDown) {
            return;
        }

        shuttingDown = true;

        logger.info("Shutting down executor service");

        shutdownProcesses();

        executorService.shutdown();

        try {
            if (!executorService.awaitTermination(10, TimeUnit.SECONDS)) {
                logger.warn("Forcing shutdown of executor service...");
                executorService.shutdownNow();
            }
        } catch (InterruptedException e) {
            logger.warn("Shutdown interrupted. Forcing shutdown...");
            executorService.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    /**
     * Shutdown all running processes
     */
    private static synchronized void shutdownProcesses() {
        logger.info("Shutting down running processes");

        for (Process process : runningProcesses) {
            try {
                // Primeiro, tentamos destruir o processo de forma normal
                process.destroy();
                if (!process.waitFor(5, TimeUnit.SECONDS)) {
                    // Se o processo não terminar dentro do limite de tempo, forçamos a
                    // destruição
                    logger.warn("Process did not terminate in time. Forcing shutdown...");
                    process.destroyForcibly();
                }
                logger.info("Process terminated: {}", process);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                // Se a espera pelo processo for interrompida, forçamos imediatamente
                logger.warn("Process interrupted. Forcing shutdown...");
                process.destroyForcibly();
            } catch (Exception e) {
                logger.warn("Error during process shutdown: {}", e.getMessage());
            }
        }

        logger.info("All processes terminated");

        runningProcesses.clear();
    }

    /**
     * Register a process to be managed by the APIUtils class
     *
     * @param process The process to register
     */
    private static synchronized void registerProcess(Process process) {
        if (shuttingDown) {
            throw new MoinexException.ApplicationShuttingDownException(
                    "Application is shutting down");
        }

        runningProcesses.add(process);
    }

    /**
     * Remove a process from the list of running processes
     *
     * @param process The process to remove
     */
    private static synchronized void removeProcess(Process process) {
        if (shuttingDown) {
            throw new MoinexException.ApplicationShuttingDownException(
                    "Application is shutting down");
        }

        runningProcesses.remove(process);
    }

    /**
     * Fetch stock prices asynchronously from an external API
     *
     * @param symbols Array of stock symbols
     * @return A CompletableFuture containing the JSON response
     */
    public static CompletableFuture<JSONObject> fetchStockPricesAsync(String[] symbols) {
        return CompletableFuture.supplyAsync(
                () -> runPythonScript(Constants.GET_STOCK_PRICE_SCRIPT, symbols), executorService);
    }

    /**
     * Fetch brazilian market indicators asynchronously from an external API
     *
     * @return A CompletableFuture containing the JSON response
     */
    public static CompletableFuture<JSONObject> fetchBrazilianMarketIndicatorsAsync() {
        return CompletableFuture.supplyAsync(
                () ->
                        runPythonScript(
                                Constants.GET_BRAZILIAN_MARKET_INDICATORS_SCRIPT, new String[0]),
                executorService);
    }

    /**
     * Execute a Python script with arguments
     *
     * @param script The script to run
     * @param args   The arguments to pass to the script
     */
    public static JSONObject runPythonScript(String script, String[] args) {
        try (InputStream scriptInputStream =
                APIUtils.class.getResourceAsStream(Constants.SCRIPT_PATH + script)) {
            if (scriptInputStream == null) {
                throw new MoinexException.ScriptNotFoundException(
                        "Python " + script + " script not found");
            }

            if (shuttingDown) {
                throw new MoinexException.ApplicationShuttingDownException(
                        "Application is shutting down");
            }

            // Script name without extension
            String scriptName = script.substring(0, script.lastIndexOf('.'));
            Path tempDirectory = Paths.get(System.getProperty("java.io.tmpdir"));
            Path tempFile = Files.createTempFile(tempDirectory, scriptName, ".py");

            setSecurePermissions(tempFile);

            Files.copy(scriptInputStream, tempFile, StandardCopyOption.REPLACE_EXISTING);

            List<String> commandList = new ArrayList<>();
            commandList.add(Constants.PYTHON_INTERPRETER);
            commandList.add(tempFile.toString());
            commandList.addAll(Arrays.asList(args));

            String[] command = commandList.toArray(new String[0]);

            logger.info("Running Python script as: {}", String.join(" ", command));

            ProcessBuilder processBuilder = new ProcessBuilder(command);
            Process process = processBuilder.start();
            registerProcess(process);

            try (BufferedReader reader =
                    new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String output = reader.lines().collect(Collectors.joining());
                int exitCode = process.waitFor();

                if (exitCode != 0) {
                    throw new MoinexException.APIFetchException(
                            "Error executing Python script. Exit code: " + exitCode);
                }

                JSONObject jsonObject = new JSONObject(output);

                logger.info("Script {} run successfully", script);
                logger.info("Output: {}", jsonObject);

                return jsonObject;
            } finally {
                synchronized (runningProcesses) {
                    removeProcess(process);
                }
            }
        } catch (InterruptedException e) {
            // Handle the case where the thread is interrupted
            Thread.currentThread().interrupt();
            throw new MoinexException.APIFetchException(
                    "Python script execution was interrupted: " + e);
        } catch (Exception e) {
            // Handle general errors and exceptions
            throw new MoinexException.APIFetchException(
                    "Error running Python script: " + e.getMessage());
        }
    }

    /**
     * Set secure file permissions to restrict access to the owner only
     *
     * @param file The file to set permissions for
     */
    private static void setSecurePermissions(Path file) throws IOException {
        // Set file permissions to be readable and writable only by the owner
        Files.setPosixFilePermissions(file, PosixFilePermissions.fromString("rw-------"));
    }
}

File: src/main/kotlin/org/moinex/util/APIUtils.kt
/*
 * Filename: APIUtils.kt (original filename: APIUtils.java)
 * Created on: January 17, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 10/03/2026
 */

package org.moinex.util

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import org.moinex.common.extension.toBACENFormat
import org.moinex.error.MoinexException
import org.moinex.model.enums.InterestIndex
import org.moinex.model.enums.PeriodType
import org.slf4j.LoggerFactory
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.nio.file.attribute.PosixFilePermissions
import java.time.LocalDate
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlin.io.path.pathString

object APIUtils {
    private val executorService = Executors.newCachedThreadPool()
    private val runningProcesses = mutableListOf<Process>()
    private val logger = LoggerFactory.getLogger(APIUtils::class.java)

    private const val DEFAULT_SCRIPT_TIMEOUT_SECONDS = 10L

    @Volatile
    private var shuttingDown = false

    @Synchronized
    fun shutdownExecutor() {
        if (shuttingDown) return

        shuttingDown = true
        logger.info("Shutting down executor service")

        shutdownProcesses()

        executorService.shutdown()

        try {
            if (!executorService.awaitTermination(10, TimeUnit.SECONDS)) {
                logger.warn("Forcing shutdown of executor service...")
                executorService.shutdownNow()
            }
        } catch (_: InterruptedException) {
            logger.warn("Shutdown interrupted. Forcing shutdown...")
            executorService.shutdownNow()
            Thread.currentThread().interrupt()
        }
    }

    @Synchronized
    private fun shutdownProcesses() {
        logger.info("Shutting down running processes")

        runningProcesses.forEach { process ->
            runCatching {
                process.destroy()
                if (!process.waitFor(5, TimeUnit.SECONDS)) {
                    logger.warn("Process did not terminate in time. Forcing shutdown...")
                    process.destroyForcibly()
                }
                logger.info("Process terminated: $process")
            }.onFailure { e ->
                when (e) {
                    is InterruptedException -> {
                        Thread.currentThread().interrupt()
                        logger.warn("Process interrupted. Forcing shutdown...")
                        process.destroyForcibly()
                    }
                    else -> logger.warn("Error during process shutdown: ${e.message}")
                }
            }
        }

        logger.info("All processes terminated")
        runningProcesses.clear()
    }

    @Synchronized
    private fun registerProcess(process: Process) {
        if (shuttingDown) {
            throw MoinexException.ApplicationShuttingDownException("Application is shutting down")
        }
        runningProcesses.add(process)
    }

    @Synchronized
    private fun removeProcess(process: Process) {
        if (shuttingDown) {
            throw MoinexException.ApplicationShuttingDownException("Application is shutting down")
        }
        runningProcesses.remove(process)
    }

    suspend fun fetchStockPrices(
        symbols: Array<String>,
        dispatcher: CoroutineDispatcher = Dispatchers.IO,
    ): JSONObject =
        withContext(dispatcher) {
            runPythonScript(Constants.GET_STOCK_PRICE_SCRIPT, symbols)
        }

    suspend fun fetchBrazilianMarketIndicators(dispatcher: CoroutineDispatcher = Dispatchers.IO): JSONObject =
        withContext(dispatcher) {
            runPythonScript(Constants.GET_BRAZILIAN_MARKET_INDICATORS_SCRIPT, emptyArray())
        }

    suspend fun fetchFundamentalAnalysis(
        symbol: String,
        period: PeriodType,
        dispatcher: CoroutineDispatcher = Dispatchers.IO,
    ): JSONObject =
        withContext(dispatcher) {
            runPythonScript(
                Constants.GET_FUNDAMENTAL_DATA_SCRIPT,
                arrayOf(symbol, "--period", period.name.lowercase(), "--format", "json"),
            )
        }

    suspend fun fetchMarketIndicatorHistory(
        indicatorType: InterestIndex,
        startDate: LocalDate,
        endDate: LocalDate,
        dispatcher: CoroutineDispatcher = Dispatchers.IO,
    ): JSONObject =
        withContext(dispatcher) {
            runPythonScript(
                Constants.GET_MARKET_INDICATOR_HISTORY_SCRIPT,
                arrayOf(
                    indicatorType.name,
                    startDate.toBACENFormat(),
                    endDate.toBACENFormat(),
                ),
            )
        }

    suspend fun fetchStockLogos(
        websites: Array<String>,
        dispatcher: CoroutineDispatcher = Dispatchers.IO,
    ): JSONObject =
        withContext(dispatcher) {
            runPythonScript(Constants.GET_STOCK_LOGO_SCRIPT, websites)
        }

    suspend fun fetchStockPriceHistory(
        symbol: String,
        startDate: String,
        endDate: String,
        specificDates: List<String>? = null,
        dispatcher: CoroutineDispatcher = Dispatchers.IO,
    ): JSONObject =
        withContext(dispatcher) {
            val args =
                if (specificDates.isNullOrEmpty()) {
                    arrayOf(symbol, startDate, endDate)
                } else {
                    arrayOf(symbol, startDate, endDate, JSONArray(specificDates).toString())
                }
            runPythonScript("get_stock_price_history.py", args)
        }

    @Deprecated("Use suspend function fetchStockPrices instead", ReplaceWith("fetchStockPrices(symbols)"))
    fun fetchStockPricesAsync(symbols: Array<String>): CompletableFuture<JSONObject> =
        CompletableFuture.supplyAsync(
            { runPythonScript(Constants.GET_STOCK_PRICE_SCRIPT, symbols) },
            executorService,
        )

    @Deprecated("Use suspend function fetchBrazilianMarketIndicators instead", ReplaceWith("fetchBrazilianMarketIndicators()"))
    fun fetchBrazilianMarketIndicatorsAsync(): CompletableFuture<JSONObject> =
        CompletableFuture.supplyAsync(
            { runPythonScript(Constants.GET_BRAZILIAN_MARKET_INDICATORS_SCRIPT, emptyArray()) },
            executorService,
        )

    @Deprecated("Use suspend function fetchFundamentalAnalysis instead", ReplaceWith("fetchFundamentalAnalysis(symbol, period)"))
    fun fetchFundamentalAnalysisAsync(
        symbol: String,
        period: PeriodType,
    ): CompletableFuture<JSONObject> =
        CompletableFuture.supplyAsync(
            {
                runPythonScript(
                    Constants.GET_FUNDAMENTAL_DATA_SCRIPT,
                    arrayOf(symbol, "--period", period.name.lowercase(), "--format", "json"),
                )
            },
            executorService,
        )

    @Deprecated("Use suspend function fetchStockLogos instead", ReplaceWith("fetchStockLogos(websites)"))
    fun fetchStockLogosAsync(websites: Array<String>): CompletableFuture<JSONObject> =
        CompletableFuture.supplyAsync(
            { runPythonScript(Constants.GET_STOCK_LOGO_SCRIPT, websites) },
            executorService,
        )

    @Deprecated(
        "Use suspend function fetchStockPriceHistory instead",
        ReplaceWith("fetchStockPriceHistory(symbol, startDate, endDate, specificDates)"),
    )
    fun fetchStockPriceHistoryAsync(
        symbol: String,
        startDate: String,
        endDate: String,
        specificDates: List<String>? = null,
    ): CompletableFuture<JSONObject> =
        CompletableFuture.supplyAsync(
            {
                val args =
                    if (specificDates.isNullOrEmpty()) {
                        arrayOf(symbol, startDate, endDate)
                    } else {
                        arrayOf(symbol, startDate, endDate, JSONArray(specificDates).toString())
                    }
                runPythonScript("get_stock_price_history.py", args)
            },
            executorService,
        )

    @Deprecated("Use suspend function with runPythonScript directly", ReplaceWith("runPythonScript(script, args)"))
    fun runPythonScriptAsync(
        script: String,
        args: Array<String>,
    ): CompletableFuture<JSONObject> =
        CompletableFuture.supplyAsync(
            { runPythonScript(script, args) },
            executorService,
        )

    fun runPythonScript(
        script: String,
        args: Array<String>,
        timeoutSeconds: Long = DEFAULT_SCRIPT_TIMEOUT_SECONDS,
    ): JSONObject {
        val scriptInputStream: InputStream =
            APIUtils::class.java.getResourceAsStream(Constants.SCRIPT_PATH + script)
                ?: throw MoinexException.ScriptNotFoundException("Python $script script not found")

        if (shuttingDown) {
            throw MoinexException.ApplicationShuttingDownException("Application is shutting down")
        }

        return runCatching {
            val scriptName = script.substringBeforeLast('.')
            val tempDirectory = Paths.get(System.getProperty("java.io.tmpdir"))
            val tempFile = Files.createTempFile(tempDirectory, scriptName, ".py")

            setSecurePermissions(tempFile)

            Files.copy(scriptInputStream, tempFile, StandardCopyOption.REPLACE_EXISTING)

            val command =
                buildList {
                    add(Constants.PYTHON_INTERPRETER)
                    add(tempFile.pathString)
                    addAll(args)
                }.toTypedArray()

            logger.info("Running Python script as: ${command.joinToString(" ")}")

            val process = ProcessBuilder(*command).start()
            registerProcess(process)

            try {
                process.inputStream.bufferedReader().use { reader ->
                    process.errorStream.bufferedReader().use { errorReader ->
                        val completed = process.waitFor(timeoutSeconds, TimeUnit.SECONDS)

                        if (!completed) {
                            logger.error("Python script '$script' timed out after $timeoutSeconds seconds")
                            process.destroyForcibly()
                            throw MoinexException.APIFetchException(
                                "Python script '$script' execution timed out after $timeoutSeconds seconds",
                            )
                        }

                        val output = reader.readText()
                        val errorOutput = errorReader.readText()
                        val exitCode = process.exitValue()

                        if (exitCode != 0) {
                            logger.error("Python script '$script' failed with exit code: $exitCode")

                            if (errorOutput.isNotEmpty()) {
                                logger.debug("Error output: $errorOutput")
                            }

                            throw MoinexException.APIFetchException(
                                "Error executing '$script'. Exit code: $exitCode",
                            )
                        }

                        val jsonObject = JSONObject(output)

                        logger.info("Script $script run successfully")
                        logger.info("Output: $jsonObject")

                        jsonObject
                    }
                }
            } finally {
                synchronized(runningProcesses) {
                    removeProcess(process)
                }
            }
        }.getOrElse { e ->
            when (e) {
                is InterruptedException -> {
                    Thread.currentThread().interrupt()
                    throw MoinexException.APIFetchException("Python script execution was interrupted: $e")
                }
                else -> throw MoinexException.APIFetchException("Error running Python script: ${e.message}")
            }
        }
    }

    private fun setSecurePermissions(file: Path) {
        val os = System.getProperty("os.name").lowercase()

        if (!os.contains("win")) {
            Files.setPosixFilePermissions(file, PosixFilePermissions.fromString("rw-------"))
        }
    }
}
